File: 2021/archived/Day04/Day4.scala
import scala.io.Source

import scala.util.chaining._
import scala.language.implicitConversions
import scala.annotation.tailrec

@main def day4 =
    type Board = Array[Array[(Int, Boolean, Int)]]
    type Row = Array[(Int, Boolean, Int)]
    type Col = Row

    val numbers = Source
        .fromFile("call.txt")
        .getLines
        .next
        .split(',')
        .map(_.toInt)
        .toList

    val boards = Source
        .fromFile("boards.txt")
        .getLines
        .filter(!_.isBlank)
        .grouped(5)
        .map(group =>
            group
                .map(
                  _.split(" +")
                      .filter(_.trim != "")
                      .map(n => (n.toInt, false, 0))
                )
                .filter(!_.isEmpty)
                .toArray
        )
        .toList

    def col(board: Board, index: Int) = board.map(_(index))
    def row(board: Board, index: Int) = board(index)

    def isWin(rowOrCol: Row | Col) =
        rowOrCol
            .forall(_._2)

    def isBoardWin(board: Board) =
        Range(0, 5, 1)
            .map(index =>
                col(board, index).pipe(isWin) || row(board, index).pipe(isWin)
            )
            .exists(identity)

    def winner(boards: List[Board]) =
        boards.find(isBoardWin) match
            case None => (boards, None)
            case win  => (boards.filter(!isBoardWin(_)), win)

    def mark(boards: List[Board], num: Int, time: Int) =
        for board <- boards
        yield for row <- board
        yield for cell <- row
        yield
            val (n, c, t) = cell
            if n == num then (n, true, time)
            else cell

    def sum(board: Board) =
        val lastCall =
            def checkWin(board: Board, direction: (Board, Int) => Row | Col) =
                Range(0, 5, 1)
                    .map(direction(board, _))
                    .find(isWin)
                    .map(_.maxBy(_._3))
                    .map(_._1)

            checkWin(board, row)
                .orElse(checkWin(board, col))
                .get

        (for
            row <- board
            cell <- row
        yield
            val (n, c, _) = cell
            if c then 0
            else n).sum * lastCall

    def firstPart(numbers: List[Int], boards: List[Board]) =
        @tailrec
        def loop(
            numbers: List[Int],
            time: Int,
            boards: List[Board],
            win: Option[Board]
        ): Board =
            win match
                case Some(win) => win
                case None =>
                    numbers match
                        case Nil => sys.error("No one win!")
                        case n :: ns =>
                            val (nextBoars, win) =
                                mark(boards, n, time).pipe(winner)
                            loop(ns, time + 1, nextBoars, win)

        loop(numbers, 1, boards, None).pipe(sum)

    def secondPart(numbers: List[Int], boards: List[Board]) =
        @tailrec
        def loop(
            numbers: List[Int],
            time: Int,
            boards: List[Board],
            lastWin: Option[Board]
        ): Board =
            boards match
                case Nil => lastWin.get
                case _ =>
                    numbers match
                        case Nil => lastWin.get
                        case n :: ns =>
                            val (nextBoars, win) =
                                mark(boards, n, time).pipe(winner)

                            loop(ns, time + 1, nextBoars, win.orElse(lastWin))

        loop(numbers, 1, boards, None).pipe(sum)

    println(s"Part 1: ${firstPart(numbers, boards)}")
    println(s"Part 1: ${secondPart(numbers, boards)}")

File: 2021/archived/Day04/Day4.fs
module Day4

open System.IO

let inline (^||^) f g a = f a || g a

type SingleBoard = (int * bool * int) [,]
type Board = SingleBoard List

let slice (boards: (int * bool * int) array array) index = boards.[index..index + 4] |> array2D
let row (board: SingleBoard) i = board.[i, *]
let col (board: SingleBoard) i = board.[*, i]

let numbers =
    (File.ReadAllLines "call.txt" |> Array.head)
        .Split [| ',' |]
    |> Array.map int
    |> Array.toList

let boards =
    let raw =
        File.ReadAllLines "boards.txt"
        |> Array.map (fun s ->
            s.Split [| ' ' |]
            |> Array.filter (fun c -> not <| (c.Trim() = ""))
            |> Array.map (fun s -> (s |> int, false, 0)))
        |> Array.filter (not << Array.isEmpty)

    let boardsNum = (raw |> Array.length) / 5

    [| for i in 0 .. boardsNum - 1 -> i * 5 |]
    |> Array.map (slice raw)
    |> Array.toList

let isWin rowOrCol =
    rowOrCol
    |> Array.forall (fun (_, selected, _) -> selected)

let isBoardWin board =
    [| 0 .. 4 |]
    |> Array.map ((row board >> isWin) ^||^ (col board >> isWin))
    |> Array.exists id

let winner boards =
    match boards |> List.tryFind isBoardWin with
    | None -> (boards, None)
    | win -> (boards |> List.filter (not << isBoardWin), win)

let mark boards num time =
    boards
    |> List.map (fun board ->
        board
        |> Array2D.map (fun (n, s, t) ->
            if n = num then
                (n, true, time)
            else
                (n, s, t)))

let sum (board: SingleBoard) =
    let lastCall =
        let checkWin (board: SingleBoard) direction =
            [| 0 .. 4 |]
            |> Array.map (direction board)
            |> Array.tryFind isWin
            |> Option.map (Array.maxBy (fun (_, _, time) -> time))
            |> Option.map (fun (n, _, _) -> n)

        checkWin board row
        |> Option.orElseWith (fun () -> checkWin board col)
        |> Option.get

    (board
     |> Array2D.map (fun (n, c, _) -> if c then 0 else n)
     |> Seq.cast<int>
     |> Seq.sum)
    * lastCall

let fisrtPart numbers boards =
    let rec loop numbers time boards win =
        match win with
        | None ->
            match numbers with
            | [] -> failwith "No one win!"
            | x :: xs ->
                mark boards x time
                |> winner
                ||> loop xs (time + 1)
        | Some (win) -> win

    loop numbers 1 boards None |> sum

let secondPart numbers boards =
    let rec loop numbers time boards lastWin =
        match boards with
        | [] -> lastWin |> Option.get
        | _ ->
            match numbers with
            | [] -> lastWin |> Option.get
            | x :: xs ->
                mark boards x time
                |> winner
                ||> fun boards win -> (boards, win |> Option.orElse lastWin)
                ||> loop xs (time + 1)

    loop numbers 1 boards None |> sum

fisrtPart numbers boards
|> printfn "First part: %d"

secondPart numbers boards
|> printfn "Second part: %d"
